File: 1717.maximum-score-from-removing-substrings.py
#
# @lc app=leetcode id=1717 lang=python3
#
# [1717] Maximum Score From Removing Substrings
#

# @lc code=start
class Solution:
    def maximumGain(self, s: str, x: int, y: int) -> int:
        stack_one = []
        stack_two = []
        ans = 0

        #two stacks and greedy     
        #x: ab y: ba

        if x > y:
            for c in s:
                #remove all ab first 
                if stack_one and c == 'b' and stack_one[-1] == 'a':
                    ans += x
                    stack_one.pop()
                else:
                    stack_one.append(c)
            s2 = ''.join(stack_one)
            for c in s2:
                if stack_two and c == 'a' and stack_two[-1] == 'b':
                    ans += y
                    stack_two.pop()
                else:
                    stack_two.append(c)
        else:
            for c in s:
                if stack_two and c == 'a' and stack_two[-1] == 'b':
                    ans += y
                    stack_two.pop()
                else:
                    stack_two.append(c)
            s2 = ''.join(stack_two)
            for c in s2:
                if stack_one and c == 'b' and stack_one[-1] == 'a':
                    ans += x
                    stack_one.pop()
                else:
                    stack_one.append(c)
        return ans
            

# @lc code=end

File: 1717.maximum-score-from-removing-substrings.java
/*
 * @lc app=leetcode id=1717 lang=java
 *
 * [1717] Maximum Score From Removing Substrings
 */

// @lc code=start
class Solution {
    public int maximumGain(String s, int x, int y) {
        // s = cdbcbbaaabab
        // s = cdbcbbaaab ans += 4
        // s = cdbcbbaa   ans += 4
        // s = cdbcba     ans += 5
        // s = cdbc       ans += 5 ans = 18
        // Correct pattern
        // s = cdbcbbaaabab
        // s = cdbcbbaaab ans += 5
        // s = cdbcbbaa   ans += 4
        // s = cdbcba     ans += 5
        // s = cdbc       ans += 5 ans = 19
        // The point here is that we should prioritize the one with the larger score
        // s = aabbaaxybbaabb, x = 5, y = 4 remove as many ab as possible
        // aaxybb 4 * 5 ans = 20

        //todo ugly but works
        int ans = 0;
        StringBuilder sb_one = new StringBuilder();
        StringBuilder sb_two = new StringBuilder();
        if (x > y){
            for (char c : s.toCharArray()){
                int len = sb_one.length();
                if(!sb_one.isEmpty() && c == 'b' && sb_one.charAt(len - 1) == 'a'){
                    ans += x;
                    sb_one.deleteCharAt(len -1);
                }else sb_one.append(c);
            }
            for(char c: sb_one.toString().toCharArray()){
                int len = sb_two.length();
                if(!sb_two.isEmpty() && c == 'a' && sb_two.charAt(len - 1) == 'b'){
                    ans += y;
                    sb_two.deleteCharAt(len -1);
                }else sb_two.append(c);
            }

        }else{
            for (char c : s.toCharArray()){
                int len = sb_one.length();
                if(!sb_one.isEmpty() && c == 'a' && sb_one.charAt(len - 1) == 'b'){
                    ans += y;
                    sb_one.deleteCharAt(len -1);
                }else sb_one.append(c);
            }
            for(char c: sb_one.toString().toCharArray()){
                int len = sb_two.length();
                if(!sb_two.isEmpty() && c == 'b' && sb_two.charAt(len - 1) == 'a'){
                    ans += x;
                    sb_two.deleteCharAt(len -1);
                }else sb_two.append(c);
            }
        }
        return ans;
    }
}
// @lc code=end

